R/defs.bzl

```diff
@@ -711,3 +715,128 @@ def r_package_with_test(pkg_name, pkg_srcs, pkg_deps, pkg_suggested_deps=[], tes
         pkg = pkg_name,
         suggested_deps = pkg_suggested_deps,
     )
+
+load(":runfiles_commands.bzl", "runfiles_commands")
+
+def _r_binary_impl(ctx):
+    if len(ctx.files.srcs) > 1:
+        fail("only one srcs file is allowed; got " + len(ctx.files.srcs))
+
+    ldeps = _library_deps(ctx.attr.deps, path_prefix=ctx.workspace_name + "/")
+
+    tools = _executables(ctx.attr.tools)
+    for dep in ctx.attr.deps:
+        tools += dep[RPackage].transitive_tools
+
+    script = "\n".join([
+        "#!/bin/bash",
+        "set -euo pipefail",
+        "",
+    ] + _env_vars(ctx.attr.env_vars) + runfiles_commands() + [("\n".join([
+        "",
+        "export PATH=\"${{PATH}}:{0}\"",
+        "",
+        "export R_LIBS_USER=$(mktemp -d)",
+        "pushd ${{RUNFILES}} > /dev/null",
+        ldeps["symlinked_library_command"],
+        "",
+        "pushd {2} > /dev/null",
+        "cleanup() {{",
+        "  set -e",
+        "  popd > /dev/null",
+        "  rm -rf ${{R_LIBS_USER}}",
+        "}}",
+        "",
+        "set +e",
+        _Rscript + "{3} {1} \"$@\"",
+        "code=$?",
+        "if [ $code -ne 0 ]; then",
+        "  cleanup",
+        "  echo \"Rscript error code: $code\"",
+        "  exit $code",
+        "fi",
+        "",
+        "cleanup",
+    ])).format(
+        _path(tools), ctx.files.srcs[0].short_path, ctx.workspace_name,
+        ctx.attr.rscript_args)])
+
+    ctx.actions.write(output = ctx.outputs.executable, content = script)
+
+    runfiles = ctx.runfiles(
+        files = ldeps["lib_files"] + ctx.files.srcs + ctx.files.data,
+        transitive_files = tools)
+    return [DefaultInfo(runfiles = runfiles)]
+
+r_binary = rule(
+    attrs = {
+        "srcs": attr.label_list(
+            # https://cran.r-project.org/doc/manuals/r-release/R-exts.html
+            allow_files = FileType([
+                ".R",
+                ".r",
+                ".S",
+                ".s",
+                ".q",
+            ]),
+            mandatory = True,
+            doc = "Single R script file to run",
+        ),
+        "deps": attr.label_list(
+            providers = [RPackage],
+            doc = "R package dependencies of type r_pkg",
+        ),
+        "data": attr.label_list(allow_files = True),
+        "env_vars": attr.string_dict(
+            doc =
+                "Extra environment variables to define before running the script",
+        ),
+        "tools": attr.label_list(
+            doc = "Executables to be made available to the script",
+        ),
+        "rscript_args": attr.string(
+            default = "--default-packages=methods,utils,grDevices,graphics",
```

Comment:
Interesting, apparently the interface for Rscript changed (at least in 3.4.3), because when I started using --default-packages, those packages were not loaded by Rscript, but loaded by Rstudio, so not having this would actually cause confusion.  I'll remove that and anyway I like the idea of a shebang.